src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,143 @@
 
     @Override
     public void addAll(List<T> list) {
-
+        int arrayLength = list.size();
+        Object[] listArray = new Object[arrayLength];
+        if (arrayLength == 0) {
+            return;
+        }
+        for (int i = 0; i < arrayLength; i++) {
+            listArray[i] = list.get(i);
+        }
+        while (arrayLength > elementData.length - size) {
+            grow();
+        }
+        System.arraycopy(listArray, 0, elementData, size, arrayLength);
+        size += arrayLength;
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        return (T) elementData[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        checkIndex(index);
+        elementData[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        T oldValue = divideAndRemoveValue(index);
+        size--;
+        return oldValue;
+
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        int elementIndex;
+        for (int i = 0; i < elementData.length; i++) {
+            if ((elementData[i] != null && elementData[i].equals(element))
+                    || elementData[i] == null && element == null) {
+                elementIndex = i;
+                size--;
+                return divideAndRemoveValue(elementIndex);
+            }
+        }
+        throw new NoSuchElementException();
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    public void checkIndexToAdd(int index) throws ArrayListIndexOutOfBoundsException {
+        if (index > size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Index " + index + " is not valid");
+        }
+    }
+
+    public void checkIndex(int index) throws ArrayListIndexOutOfBoundsException {
+        if (index < 0 || index >= size) {
+            throw new ArrayListIndexOutOfBoundsException("Index " + index + " is not valid");
+        }
+    }
+
+    private void growIfArrayFull() {
+        if (size == elementData.length) {
+            grow();
+        }
+    }
+
+    private void grow() {
+        int newCapacity = elementData.length + (elementData.length >> 1);
+        Object[] copiedArray = new Object[newCapacity];
+        System.arraycopy(elementData, 0, copiedArray, 0, size);
+        elementData = copiedArray;
+    }
+
+    private void divideAndSetNewValue(T value, int index) {
+        System.arraycopy(elementData, index,
+                elementData, index + 1,
+                size - index);
+        elementData[index] = value;
+    }
+
+    private T divideAndRemoveValue(int index) {
+        T oldValue = (T) elementData[index];
+        if (index == size) {
+            System.arraycopy(elementData, 0, elementData, 0, size);
```

Comment:
The condition `if (index == size)` is incorrect. The index should never be equal to the size when removing an element, as it would be out of bounds. This check should be removed or corrected to prevent incorrect behavior.